_posts/2020-06-13-split-your-rspec-individual-test-examples-between-github-action-nodes.md

```diff
@@ -0,0 +1,126 @@
+
+## Consider the split
+
+Imagine you have a project with 30 RSpec spec files. Each file contains multiple test examples (RSpec "`it`s"). Most of the files are fairly robust, fast unit tests. Let's say there are also some slower files, like feature specs. Perhaps one such feature spec file takes approximately ~5 minutes to execute.
+
+When we run different spec files on different parallel machines, we strive for similar execution time on all of them. In a described scenario, even if we run 30 parallel jobs (each one running just one test file), the 5 minute feature spec would be a bottleneck of the whole build. 29 machines may finish their work in a matter of seconds, but the whole build won't be complete until the 1 remaining node finishes executing its file.
+
+## Divide and conquer
+
+To solve the problem of a slow test file, we need to split what's inside it. We could refactor it and ensure the test examples live in separate, smaller test files. There are two problems with that though:
+
+First, it needs work. Although admittedly quite plausible in a described scenario, in real life it's usually not just the one file that's causing problems. Oftentimes there is a number of slow and convoluted test files, with their own complex setups, like nested `before` blocks, `let`s, etc. We've all seen them (and probably contributed to them ending-up this way), haven't we? ;-) Refactoring files like that is no fun, and there seem to always be more higher prio work to be done, at least from our experience.
+
+Second, we belive that the code organization should be based on other considerations. How you create your files and classes is most likely a result of following some approach agreed upon in your project. Dividing classes into smaller ones so that the CI build can run faster encroaches on your conventions. It might be more disturbing to some than the others, but we feel it's fair to say it'd be best to avoid - if there was a better way to achieve the same...
+
+## Enter split by test examples
+
+As you certainly know, RSpec allows us to run individual examples instead of whole files. We decided to take advantage of that, and solve the problem of bottleneck test files by gathering information about individual examples from such slower files. Such examples are then dynamically distributed between your parallel nodes and run individually, so no individual file can be a bottleneck for the whole build. What's important, no additional work is needed - this is done automatically by the gem. Each example is run in its correct context that's set-up exactly the same as if you had run the whole file.
```

Comment:
` by the gem.` => ` by the knapsack_pro gem`. It wasn't obvious what kind of gem you meant.